### PR TITLE
Question Service Categories Id

### PR DESCRIPTION
**Pull request by elaineshijie**
### Description
- Added categoriesId and meta field in questions
- Added descriptionText and descriptionHtml to description for validation
- Added questions.json file to store all questions to add (description in description object is temporary and testCases are not checked yet)
- **Note**: Please modify your `.env` file accordingly to use `question-service-D5` database. `.env` should have 
```
DEV_PORT = 8003
DEV_URI = mongodb+srv://<username>:<password>@main.24fab.mongodb.net/question-service-D5?retryWrites=true&w=majority&appName=main
ACCESS_TOKEN_SECRET=<copy from previous file>
```
- Updating of question details of deleted question not allowed (soft delete) 
- GET /filter and /random can only take in categoriesId as params 
- Yet to populate questions into `question-service-D5` database!!

### Additional information
- All (POST, PUT, GET(/filter, /random) will only take in categoriesId but not categories
-  categoriesIdToCategories = {
    0 : "ARRAYS",
    1 : "ALGORITHMS",
    2 : "DATABASES",
    3 : "DATA STRUCTURES",
    4 : "BRAINTEASER",
    5 : "STRINGS",
    6 : "BIT MANIPULATION",
    7 : "RECURSION"
  };
- Add unauthorised error
- https://peerprep-3564.postman.co/workspace/Peerprep-Workspace~7bfec49a-e7b4-4c84-b244-6ecc2b112326/collection/38418833-70dddbb1-9be6-4d98-a599-9e116f908375?action=share&creator=38418833 
- `meta: {
 publicTestCaseCount: Number,
 privateTestCaseCount: Number,
 totalTestCaseCount: Number 
}`
- `description: {
descriptionText: String,
descriptionHtml: String
}`

### CODE REVIEW

File: question-service/controller/question-controller.js [hunk 1]
```diff
@@ -41,11 +44,29 @@ const createQuestion = async (req, res, next) => {
...
+    // count number of isPublic test cases, !isPublic test cases, and total test cases  
+    const metaData = {
+      publicTestCaseCount: req.body.testCases.filter(testCase => testCase.isPublic).length,
+      privateTestCaseCount: req.body.testCases.filter(testCase => !testCase.isPublic).length,
+      totalTestCaseCount: req.body.testCases.length
+    }
+
+    const newQuestion = {
+      ...req.body,
+      categories: categoriesString,
```
[the remark this answers is not in the document]
Comment (elaineshijie) [reply question-service/controller/question-controller.js:60]: Currently when creating new Questions, both categoriesId and categories(as string) are stored. Would you like me to remove categories then?

File: question-service/controller/question-controller.js [hunk 2]
```diff
@@ -179,20 +214,44 @@ const updateQuestionById = async (req, res, next) => {
...
+        (id) => categoriesIdToCategories[id]
+      );
+
+      updatedQuestionDetails = {
+        ...req.body,
+        categories: categoriesString,
+      };
+    }
+
+    if (testCases) {
```
[the remark this answers is not in the document]
Comment (elaineshijie) [reply question-service/controller/question-controller.js:230]: it is not always true, updateQuestions do not require all data from questions (so there would be a case where users do not update testcases) unless frontend always send back question with all fields.

[the remark this answers is not in the document]
Comment (elaineshijie) [reply question-service/controller/question-controller.js:230]: testCases and meta will always exist due to validation done in creating questions (POST). If updating questions do not specify testCases, then there will be no change to testCases and meta in database.

File: question-service/controller/question-controller.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (elaineshijie) [reply question-service/controller/question-controller.js:60]: Clarification: will remove storing of categories string and return an additional field categories based on categoriesId on data instead

**Review by HollaG — APPROVED** [review]

**Pull request merged**